entities/src/main/java/org/odk/collect/entities/javarosa/filter/PullDataFunctionHandler.kt

```diff
@@ -0,0 +1,49 @@
+package org.odk.collect.entities.javarosa.filter
+
+import org.javarosa.core.model.condition.EvaluationContext
+import org.javarosa.core.model.condition.IFunctionHandler
+import org.javarosa.xpath.expr.XPathFuncExpr
+import org.odk.collect.entities.javarosa.intance.LocalEntitiesInstanceAdapter
+import org.odk.collect.entities.storage.EntitiesRepository
+
+class PullDataFunctionHandler(
+    entitiesRepository: EntitiesRepository,
+    private val fallback: IFunctionHandler? = null
+) : IFunctionHandler {
+
+    private val instanceAdapter = LocalEntitiesInstanceAdapter(entitiesRepository)
+
+    override fun getName(): String {
+        return NAME
+    }
+
+    override fun getPrototypes(): List<Array<Class<Any>>> {
+        return emptyList()
+    }
+
+    override fun rawArgs(): Boolean {
+        return true
+    }
+
+    override fun realTime(): Boolean {
+        return false
+    }
+
+    override fun eval(args: Array<Any>, ec: EvaluationContext): Any {
+        val instanceId = XPathFuncExpr.toString(args[0])
+        val child = XPathFuncExpr.toString(args[1])
+        val filterChild = XPathFuncExpr.toString(args[2])
+        val filterValue = XPathFuncExpr.toString(args[3])
+
+        return if (instanceAdapter.supportsInstance(instanceId)) {
+            instanceAdapter.queryEq(instanceId, filterChild, filterValue)!!.firstOrNull()
```

Comment:
It looks like queryEq can't return null values so we could update that function by removing `?` from the returned type and then get rid of those `!!`.